### PR TITLE
stop the loop when listener is closed

### PR DESCRIPTION
**Pull request by WeidiDeng**
Fix [258](https://github.com/mholt/caddy-l4/issues/258).

### CODE REVIEW

**Comment by mholt** [comment]
Thanks Weidi! As usual you beat me to it. Do you know why this is needed though? Like why doesn't closing the listener bump out of the Accept loop? (Only when SSH connections are active...?)

**Comment by WeidiDeng** [comment]
The listener wrapper is used in my own listener plugin, which is only used on linux. I guess because listeners on linux use port reuse by default, the error there was temporary. On other platforms, it's implemented via a deadline, which is temporary.

**Comment by WeidiDeng** [comment]
It's not ssh though, the listener loop is still running, and the error [here](https://github.com/golang/go/blob/master/src/net/http/server.go#L3334) is never received, because this listener silently discard it.

File: layer4/listener.go [hunk 1]
```diff
@@ -114,25 +116,35 @@ type listener struct {
 	logger        *zap.Logger
 	compiledRoute Handler
 
+	closed atomic.Bool
```
Comment (mholt) [inline layer4/listener.go:119]: Would a [sync.Once](https://pkg.go.dev/sync#Once) be simpler here?

**Review by mholt — COMMENTED** [review]
Thanks for working on this!

Just one idea to make things a bit simpler...

File: layer4/listener.go (hunk 1 above)
Comment (WeidiDeng) [reply layer4/listener.go:119]: `sync.Once` is used to make sure something happens once, not used to test if something happened. It doesn't fit in the struct

Comment (mholt) [reply layer4/listener.go:119]: Ah, I see below, the done channel is used for signaling.

**Comment by Monviech** [comment]
Should I test this again or is it ready?

**Comment by mholt** [comment]
I'll circle back to this soon! Been busy

**Comment by Monviech** [comment]
@mholt No stress just wanted to ask <3

**Review by mholt — APPROVED** [review]

**Comment by mholt** [comment]
CI failure seems unrelated

**Pull request merged**